Find in Page should return results in the correct order with Site Isolation Enabled iOS

colelao opened the pull request:
#### 0f4a9392b89987302460c37d4f97b0be973f3d19
<pre>
Find in Page should return results in the correct order with Site Isolation Enabled iOS
<a href="https://rdar.apple.com/165724673">rdar://165724673</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=303429">https://bugs.webkit.org/show_bug.cgi?id=303429</a>

Reviewed by Sihui Liu and Aditya Keerthi.

Currently, matches are returned in any order they are found.
This patch sorts the results by the order they appear in
the frame tree.

I added two new site isolation API tests to ensure that this
works correctly.

Test: Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm

* Source/WebKit/UIProcess/FindTextMatchesCallbackAggregator.cpp:
(WebKit::FindTextMatchCallbackAggregator::create):
(WebKit::FindTextMatchCallbackAggregator::foundMatches):
(WebKit::FindTextMatchCallbackAggregator::~FindTextMatchCallbackAggregator):
(WebKit::FindTextMatchCallbackAggregator::FindTextMatchCallbackAggregator):
* Source/WebKit/UIProcess/FindTextMatchesCallbackAggregator.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::findTextRangesForStringMatches):
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView offsetFromPosition:toPosition:]):
* Source/WebKit/WebProcess/WebPage/WebFoundTextRangeController.cpp:
(WebKit::WebFoundTextRangeController::findTextRangesForStringMatches):
* Source/WebKit/WebProcess/WebPage/WebFoundTextRangeController.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::findTextRangesForStringMatches):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm:
(TestWebKitAPI::(SiteIsolation, FindStringAcrossMultipleFramesOrderIOS)):
(TestWebKitAPI::(SiteIsolation, FindStringNestedFramesOrderIOS)):

Canonical link: <a href="https://commits.webkit.org/303929@main">https://commits.webkit.org/303929@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6da04dcdfd268646ce62fb80f439602222387ce5

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/133883 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/6394 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/45076 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/141460 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/85943 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/5886d82e-5254-4171-89ff-a4a19c98d071) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/135753 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/6924 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/6258 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/102416 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/69726 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/4aa77c04-5375-46f8-8b2a-25efcfe77df7) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/136830 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/4879 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/120021 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/83215 "Passed tests") | | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/5e703b24-ac4b-4e0b-81ee-186feff4f7cd) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/4758 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/2377 "Passed tests") | [⏳ 🛠 wpe-cairo ](https://ews-build.webkit.org/#/builders/WPE-Cairo-Build-EWS "Waiting in queue, processing has not started yet") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/113915 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/38137 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/144106 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/6064 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/38717 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/110783 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/6146 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/5166 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/110986 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/28174 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/4610 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/116276 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/59811 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/6116 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/34549 "Passed tests") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/5962 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/69580 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/6207 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/6070 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebKit/UIProcess/FindTextMatchesCallbackAggregator.cpp:
```diff
@@ -37,12 +38,35 @@ Ref<FindTextMatchCallbackAggregator> FindTextMatchCallbackAggregator::create(Com
+        bool aIsPDF = std::holds_alternative<WebFoundTextRange::PDFData>(a.data);
+        bool bIsPDF = std::holds_alternative<WebFoundTextRange::PDFData>(b.data);
+
+        // DOM always comes before PDF when in same frame
```
We don't support embedded PDFs at all on iOS (only main frame), so DOM and PDF content would never be in the same frame. (So the logic is still fine, but the comment is inaccurate)

szewai commented on Source/WebKit/WebProcess/WebPage/WebFoundTextRangeController.cpp:
```diff
@@ -77,8 +77,12 @@ void WebFoundTextRangeController::findTextRangesForStringMatches(const String& s
+    HashMap<WebCore::FrameIdentifier, uint64_t> frameOrderMap;
     uint64_t order = 0;
+    for (RefPtr frame = m_webPage->corePage()->mainFrame(); frame; frame = frame->tree().traverseNext())
+        frameOrderMap.set(frame->frameID(), order++);
```
You might need to either update this to `++order` (not use 0 as a value), or you need to use `frameOrderMap.getOptional` below, otherwise we won't be able to distinguish between key does exist v.s. the value of the key is 0.

szewai reviewed (changes requested):
We should let web process reply `HashMap<FrameIdentifier, WebFoundTextRange>` and let UI process assembles it based on the latest frame order in UI process (you could also traverse frames in UI process, see `WebPageProxy::webContentProcessesWithFrame()`).
There is a chance that different web processes have different views of frame tree (given the change is async), and we should use the one in UI process when returning the results to client (and UI process gets to filter out the ranges if frame is already removed by the time of `~FindTextMatchCallbackAggregator()`).

colelao committed:
Find in Page should return results in the correct order with Site Isolation Enabled iOS
rdar://165724673
https://bugs.webkit.org/show_bug.cgi?id=303429

Reviewed by Sihui Liu and Aditya Keerthi.

Currently, matches are returned in any order they are found.
This patch sorts the results by the order they appear in
the frame tree.

I added two new site isolation API tests to ensure that this
works correctly.

Test: Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm

* Source/WebKit/UIProcess/FindTextMatchesCallbackAggregator.cpp:
(WebKit::FindTextMatchCallbackAggregator::create):
(WebKit::FindTextMatchCallbackAggregator::foundMatches):
(WebKit::FindTextMatchCallbackAggregator::~FindTextMatchCallbackAggregator):
(WebKit::FindTextMatchCallbackAggregator::FindTextMatchCallbackAggregator):
* Source/WebKit/UIProcess/FindTextMatchesCallbackAggregator.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::findTextRangesForStringMatches):
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView offsetFromPosition:toPosition:]):
* Source/WebKit/WebProcess/WebPage/WebFoundTextRangeController.cpp:
(WebKit::WebFoundTextRangeController::findTextRangesForStringMatches):
* Source/WebKit/WebProcess/WebPage/WebFoundTextRangeController.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::findTextRangesForStringMatches):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm:
(TestWebKitAPI::(SiteIsolation, FindStringAcrossMultipleFramesOrderIOS)):
(TestWebKitAPI::(SiteIsolation, FindStringNestedFramesOrderIOS)):

Canonical link: https://commits.webkit.org/303929@main
